azplugins/flow.py

```diff
@@ -726,13 +730,28 @@ def __call__(self, timestep):
             _velocity,_ = np.histogram(x, bins=self.bins, range=self.range, weights=v)
             self._velocity += _velocity
 
+            # temperature histogram
```

Comment:
To get the temperature, it might be better to do mass-averaged histograms? Also, since we are doing a lot of histograms, maybe we should combine all of these using `numpy.digitize` and `numpy.bincount`? I think something like this would cover all properties of interest:

```py
# add self.edges in constructor
binids = np.digitize(x, self.edges)

counts = np.bincount(binids)
bin_mass = np.bincount(binids, weights=m)
vn = np.bincount(binids, weights=v)
vm = np.bincount(binids, weights=v*m)
vsqm = np.bincount(binids, weights=v**2*m)

kT = vsqm/(3*(counts-1))
```

We could also clarify properties like `number_density` (from `counts`) vs. a new property `mass_density` (from `bin_mass`). We could also clarify that `velocity` is the number average, and add a new `mass_velocity` or something like that.

For `mass_density`, `mass_velocity`, and `temperature`, would you want to compute these in each frame (and then average over frames), or accumulate between frames and then average at the end? I think both are valid.

Reply:
I would accumulate between frames and average when the user requests the data, but it makes no difference, this is just a preference.

Reply:
That's fine by me. I was worried about `<x/y>` vs. `<x>/<y>`, as I suppose this could make a difference if the number of counts in a bin is small.

Reply:
The way you calculate the temperature now only takes the velocity component in `flow_axis` into account. I think I'd prefer to calculate the temperature from all velocity components, it also looks like that this is what you intended, by dividing by 3?

Reply:
Ah yes, that's a good point. Yes, it would make sense to use all three components to get kT. Do you want to assume that the flow is a vector (*u*(x)) or that it can only have one nonzero component (uy(x))?

If you assume a vector, we could modify the `velocity` so that it returns the vector (no need to specify `flow_axis`). `np.bincount` will not take a vector as a weight, but you could call it on each component of the velocity separately:
```py
m = snap.particles.mass
v = snap.particles.velocity
for dim in range(3):
   self._velocity[dim] += np.bincount(binids,v[:,dim])
   self._mass_velocity[dim] += np.bincount(binids,m*v[:,dim])
```